File: backend/kernel/intent.py
"""Intent Declaration —— 意图 → 最小权限（P0-B）。

.. deprecated:: P0-B
    **权威合成在 Rust** ``tevarn_kernel::intent`` / host ``apply_intent``。
    ``apply_intent_to_process`` 优先走 Rust RPC；失败才用本文件本地逻辑（fallback）。
"""

from __future__ import annotations

import time
from dataclasses import dataclass, field
from typing import Any

from backend.kernel.capability import CapabilityToken

# 默认可自由授予的安全能力（只读类）
DEFAULT_GRANTABLE: frozenset[str] = frozenset({
    "file_read", "grep", "glob", "web_search", "web_extract",
    "session_search", "memory", "knowledge_search", "wiki_search",
    # 与 Rust intent::DEFAULT_GRANTABLE 对齐（主会话编制/目标/技能）
    "crew_steward", "clarify", "use_tool_pack", "current_time",
    "okr_goal", "manage_goal", "autopilot", "manage_skill", "manage_mcp",
})

# 高危能力：需 constraints.allow_risky=True 显式接受
RISKY_CAPABILITIES: frozenset[str] = frozenset({
    "terminal", "file_write", "file_edit", "browser", "computer",
    "delegate_task", "cronjob", "send_message",
})


@dataclass(frozen=True)
class IntentDeclaration:
    goal: str
    capabilities: tuple[str, ...] = ()
    constraints: dict[str, Any] = field(default_factory=dict)

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> "IntentDeclaration":
        if not isinstance(data, dict):
            raise ValueError("intent declaration 必须是 dict")
        goal = str(data.get("goal") or "").strip()
        if not goal:
            raise ValueError("intent declaration 缺少 goal")
        caps = data.get("capabilities") or []
        if not isinstance(caps, (list, tuple)):
            raise ValueError("capabilities 必须是列表")
        constraints = data.get("constraints") or {}
        if not isinstance(constraints, dict):
            raise ValueError("constraints 必须是 dict")
        return cls(
            goal=goal,
            capabilities=tuple(str(c) for c in caps),
            constraints=dict(constraints),
        )

    def to_dict(self) -> dict[str, Any]:
        return {
            "goal": self.goal,
            "capabilities": list(self.capabilities),
            "constraints": self.constraints,
        }


def synthesize_capabilities(
    intent: IntentDeclaration,
    *,
    grantable: frozenset[str] = DEFAULT_GRANTABLE,
    risky: frozenset[str] = RISKY_CAPABILITIES,
) -> tuple[list[str], list[str]]:
    """按策略合成能力集。返回 (granted, dropped)。

    规则：
    - grantable 内的能力直接授予
    - risky 内的能力仅当 constraints.allow_risky=True 时授予，否则剔除
    - 既不在 grantable 也不在 risky 的未知能力：剔除（白名单原则，宁缺毋滥）
    - 未声明任何能力：授予 grantable 全集（只读探索是安全的默认）
    """
    allow_risky = bool(intent.constraints.get("allow_risky"))
    requested = list(intent.capabilities) if intent.capabilities else sorted(grantable)

    granted: list[str] = []
    dropped: list[str] = []
    for cap in requested:
        if cap in grantable:
            granted.append(cap)
        elif cap in risky:
            (granted if allow_risky else dropped).append(cap)
        else:
            dropped.append(cap)
    return granted, dropped


def synthesize_token(
    intent: IntentDeclaration,
    *,
    parent_token: CapabilityToken | None = None,
    process_id: str = "",
    grantable: frozenset[str] = DEFAULT_GRANTABLE,
    risky: frozenset[str] = RISKY_CAPABILITIES,
) -> tuple[CapabilityToken, list[str]]:
    """声明 → 能力令牌。有父令牌时 narrow 到父集子集（可能进一步缩小 granted）。"""
    granted, dropped = synthesize_capabilities(intent, grantable=grantable, risky=risky)

    ttl = intent.constraints.get("ttl_seconds")
    expires_at = time.time() + float(ttl) if ttl else None

    if parent_token is not None:
        # 父令牌不允许的能力从 granted 移到 dropped（narrowing 会抛异常，先过滤）
        if "*" not in parent_token.capabilities:
            still: list[str] = []
            for cap in granted:
                (still if cap in parent_token.capabilities else dropped).append(cap)
            granted = still
        token = parent_token.narrow(granted, process_id=process_id, expires_at=expires_at)
        return token, dropped

    return (
        CapabilityToken(
            capabilities=frozenset(granted),
            process_id=process_id,
            expires_at=expires_at,
        ),
        dropped,
    )


def apply_intent_to_process(
    kernel: Any,
    process_id: str,
    intent: IntentDeclaration | dict[str, Any],
    *,
    parent_token: CapabilityToken | None = None,
    grantable: frozenset[str] = DEFAULT_GRANTABLE,
    risky: frozenset[str] = RISKY_CAPABILITIES,
) -> tuple[CapabilityToken, list[str]]:
    """生产路径：声明 → 合成令牌 → 挂载进程。优先 Rust host RPC。"""
    if isinstance(intent, dict):
        decl = IntentDeclaration.from_dict(intent)
    elif isinstance(intent, IntentDeclaration):
        decl = intent
    else:
        raise TypeError("intent 须为 IntentDeclaration 或 dict")

    # Rust client path
    apply_rpc = getattr(kernel, "apply_intent", None)
    if callable(apply_rpc) and hasattr(kernel, "_call"):
        try:
            return apply_rpc(process_id, decl.to_dict(), parent_token=parent_token)
        except TypeError:
            # RustAgentKernel.apply_intent signature below
            pass
        try:
            return kernel.apply_intent(  # type: ignore[misc]
                process_id, decl.to_dict(), parent_token=parent_token
            )
        except Exception:
            pass

    resolve = getattr(kernel, "_resolve_process", None)
    if resolve is None:
        resolve = getattr(kernel, "get_process", None)
    proc = resolve(process_id) if callable(resolve) else None
    if proc is None and hasattr(kernel, "get_process"):
        proc = kernel.get_process(process_id)
    if proc is None:
        raise ValueError(f"未知进程 {process_id}")

    ptok = parent_token
    if ptok is None and getattr(proc, "parent_id", None):
        parent = None
        if hasattr(kernel, "get_process"):
            parent = kernel.get_process(proc.parent_id)
        if parent is not None:
            ptok = getattr(parent, "token", None)

    token, dropped = synthesize_token(
        decl,
        parent_token=ptok,
        process_id=process_id,
        grantable=grantable,
        risky=risky,
    )
    granted = sorted(token.capabilities)
    if proc.capabilities is not None or granted:
        proc.capabilities = granted
    proc.token = token
    proc.meta = dict(proc.meta or {})
    proc.meta["intent"] = decl.to_dict()
    proc.meta["intent_dropped"] = list(dropped)
    share = getattr(kernel, "_share_process", None)
    if callable(share):
        share(proc)
    persist = getattr(kernel, "_persist_process", None)
    if callable(persist):
        try:
            persist(proc)
        except Exception:
            pass
    return token, dropped

File: crates/takton-kernel/src/intent.rs
//! Intent declaration → minimal capability synthesis (P0-B).

use std::collections::{BTreeMap, BTreeSet};
use std::time::{SystemTime, UNIX_EPOCH};

use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::capability::CapabilityToken;
use crate::error::{KernelError, KernelResult};

pub static DEFAULT_GRANTABLE: &[&str] = &[
    "file_read",
    "grep",
    "glob",
    "web_search",
    "web_extract",
    "session_search",
    "memory",
    "knowledge_search",
    "wiki_search",
    // Main-chat orchestration (tool packs always inject these; token must match)
    "crew_steward",
    "clarify",
    "use_tool_pack",
    "current_time",
];

pub static RISKY_CAPABILITIES: &[&str] = &[
    "terminal",
    "file_write",
    "file_edit",
    "browser",
    "computer",
    "delegate_task",
    "cronjob",
    "send_message",
    // abstract risky crew caps
    "command",
    "file_rw",
];

/// P1 IPC / multi-agent caps — grantable when explicitly requested (not in readonly default).
pub static IPC_CAPABILITIES: &[&str] = &[
    "ipc",
    "ipc_send",
    "ipc_recv",
    "agent_comm",
];

fn now_secs() -> f64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_secs_f64())
        .unwrap_or(0.0)
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct IntentDeclaration {
    pub goal: String,
    pub capabilities: Vec<String>,
    pub constraints: BTreeMap<String, Value>,
}

impl IntentDeclaration {
    /// Safe default: read-only exploration (grantable set).
    pub fn default_readonly(goal: &str) -> Self {
        Self {
            goal: if goal.trim().is_empty() {
                "general assistance (read-only defaults)".into()
            } else {
                goal.to_string()
            },
            capabilities: vec![], // empty → full DEFAULT_GRANTABLE
            constraints: BTreeMap::new(),
        }
    }

    pub fn from_dict(data: &Value) -> KernelResult<Self> {
        let obj = data
            .as_object()
            .ok_or_else(|| KernelError::Invalid("intent declaration 必须是 dict".into()))?;
        let goal = obj
            .get("goal")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .trim()
            .to_string();
        if goal.is_empty() {
            return Err(KernelError::Invalid("intent declaration 缺少 goal".into()));
        }
        let capabilities = obj
            .get("capabilities")
            .and_then(|v| v.as_array())
            .map(|arr| {
                arr.iter()
                    .filter_map(|x| x.as_str().map(|s| s.to_string()))
                    .collect()
            })
            .unwrap_or_default();
        let constraints = obj
            .get("constraints")
            .and_then(|v| v.as_object())
            .map(|m| m.iter().map(|(k, v)| (k.clone(), v.clone())).collect())
            .unwrap_or_default();
        Ok(Self {
            goal,
            capabilities,
            constraints,
        })
    }

    pub fn to_dict(&self) -> Value {
        serde_json::json!({
            "goal": self.goal,
            "capabilities": self.capabilities,
            "constraints": self.constraints,
        })
    }

    pub fn allow_risky(&self) -> bool {
        self.constraints
            .get("allow_risky")
            .and_then(|v| v.as_bool())
            .unwrap_or(false)
    }

    pub fn token_budget_hint(&self) -> Option<i64> {
        self.constraints
            .get("token_budget")
            .and_then(|v| v.as_i64())
            .or_else(|| {
                self.constraints
                    .get("token_budget")
                    .and_then(|v| v.as_f64().map(|f| f as i64))
            })
    }
}

pub fn synthesize_capabilities(intent: &IntentDeclaration) -> (Vec<String>, Vec<String>) {
    let grantable: BTreeSet<&str> = DEFAULT_GRANTABLE.iter().copied().collect();
    let risky: BTreeSet<&str> = RISKY_CAPABILITIES.iter().copied().collect();
    let ipc: BTreeSet<&str> = IPC_CAPABILITIES.iter().copied().collect();
    let allow_risky = intent.allow_risky();

    let requested: Vec<String> = if intent.capabilities.is_empty() {
        let mut g: Vec<_> = grantable.iter().map(|s| (*s).to_string()).collect();
        g.sort();
        g
    } else {
        intent.capabilities.clone()
    };

    let mut granted = Vec::new();
    let mut dropped = Vec::new();
    for cap in requested {
        if grantable.contains(cap.as_str()) {
            granted.push(cap);
        } else if ipc.contains(cap.as_str()) {
            // explicit request only (not in empty-default readonly set)
            granted.push(cap);
        } else if risky.contains(cap.as_str()) {
            if allow_risky {
                granted.push(cap);
            } else {
                dropped.push(cap);
            }
        } else {
            // Unknown: whitelist only — drop
            dropped.push(cap);
        }
    }
    (granted, dropped)
}

pub fn synthesize_token(
    intent: &IntentDeclaration,
    process_id: &str,
    parent: Option<&CapabilityToken>,
) -> KernelResult<(CapabilityToken, Vec<String>)> {
    let (mut granted, mut dropped) = synthesize_capabilities(intent);

    // Parent filter before narrow (avoid CapabilityEscalation panic path)
    if let Some(parent) = parent {
        if !parent.capabilities.contains("*") {
            let mut still = Vec::new();
            for cap in granted {
                if parent.allows(&cap) || parent.capabilities.contains(&cap) {
                    still.push(cap);
                } else {
                    dropped.push(cap);
                }
            }
            granted = still;
        }
    }

    let ttl = intent
        .constraints
        .get("ttl_seconds")
        .and_then(|v| v.as_f64())
        .or_else(|| {
            intent
                .constraints
                .get("ttl_seconds")
                .and_then(|v| v.as_i64().map(|i| i as f64))
        });
    let expires_at = ttl.map(|t| now_secs() + t);

    let token = if let Some(parent) = parent {
        parent.narrow(granted.iter().cloned(), process_id, expires_at)?
    } else {
        CapabilityToken::new(granted.iter().cloned(), process_id, expires_at)
    };
    Ok((token, dropped))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn default_empty_caps_is_grantable() {
        let i = IntentDeclaration::default_readonly("explore");
        let (g, d) = synthesize_capabilities(&i);
        assert!(d.is_empty());
        assert!(g.contains(&"file_read".into()));
        assert!(!g.contains(&"terminal".into()));
    }

    #[test]
    fn risky_dropped_without_flag() {
        let mut i = IntentDeclaration::default_readonly("x");
        i.capabilities = vec!["file_read".into(), "terminal".into()];
        let (g, d) = synthesize_capabilities(&i);
        assert!(g.contains(&"file_read".into()));
        assert!(d.contains(&"terminal".into()));
    }

    #[test]
    fn risky_granted_with_flag() {
        let mut i = IntentDeclaration::default_readonly("x");
        i.capabilities = vec!["terminal".into()];
        i.constraints
            .insert("allow_risky".into(), serde_json::json!(true));
        let (g, d) = synthesize_capabilities(&i);
        assert!(g.contains(&"terminal".into()));
        assert!(d.is_empty());
    }

    #[test]
    fn parent_narrows_token() {
        let parent = CapabilityToken::new(["file_read", "grep"], "p", None);
        let mut i = IntentDeclaration::default_readonly("x");
        i.capabilities = vec!["file_read".into(), "glob".into()];
        let (tok, dropped) = synthesize_token(&i, "c", Some(&parent)).unwrap();
        assert!(tok.allows("file_read"));
        assert!(!tok.allows("glob"));
        assert!(dropped.contains(&"glob".into()));
    }
}
